apiserver/internal/handlers/resources/download/downloader.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package download
+
+import (
+	"context"
+	"crypto/sha512"
+	"encoding/hex"
+	"io"
+	"os"
+
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/errors"
+)
+
+const (
+	// ErrUnexpectedHash is returned when the hash of the downloaded resource
+	// does not match the expected hash.
+	ErrUnexpectedHash = errors.ConstError("downloaded resource has unexpected hash")
+
+	// ErrUnexpectedSize is returned when the size of the downloaded resources
+	// does not match the expected size.
+	ErrUnexpectedSize = errors.ConstError("downloaded resource has unexpected size")
+)
+
+// FileSystem defines a file system for modifying files on a users system.
+type FileSystem interface {
+	// CreateTemp creates a new temporary file in the directory dir,
+	// opens the file for reading and writing, and returns the resulting file.
+	CreateTemp(dir, pattern string) (*os.File, error)
+
+	// Open opens the named file for reading. If successful, methods on
+	// the returned file can be used for reading; the associated file
+	Open(name string) (*os.File, error)
+
+	// Remove removes the named file or (empty) directory.
+	Remove(name string) error
+}
+
+// Downloader validates a resource blob.
```

Comment:
question: If the Downloader's purpose is to validate a blob, why call it a Downloader rather than BlobValidator or the like?

Reply:
I originally called it Validate, but it was pointed out (https://github.com/juju/juju/pull/18669#pullrequestreview-2582308407, https://github.com/juju/juju/pull/18669#discussion_r1937593173) that it is actually doing the downloading, and returns a ReadCloser, which a validator wouldn't do.